waspc/data/Generator/templates/server/src/auth/utils.ts

```diff
@@ -102,3 +187,64 @@ export async function validateAndGetAdditionalFields(data: {
   }
   return result;
 }
+
+export type EmailProviderData = {
+  password: string;
+  isEmailVerified: boolean;
+  emailVerificationSentAt: Date | null;
+  passwordResetSentAt: Date | null;
+}
+
+export type UsernameProviderData = {
+  password: string;
+}
+
+export type OAuthProviderData = {}
+
+export type ProviderData = {
+  email: EmailProviderData;
+  username: UsernameProviderData;
+  oauth: OAuthProviderData;
+}
+
+export function deserializeProviderData<ProviderName extends keyof ProviderData>(
+  providerData: string,
+  { shouldRemovePasswordField = false }: { shouldRemovePasswordField?: boolean } = {},
+): ProviderData[ProviderName] {
+  // NOTE: We are letting JSON.parse throw an error if the providerData is not valid JSON.
+  let data = JSON.parse(providerData) as ProviderData[ProviderName];
```

Comment:
So here we don't check if data in runtime actually has correct shape, right? I guess we could if we really wanted, but we assume it will all be fine. If we found error, the question is I gues what we would do? It would be a tricky situation, it means we have data inconsistency.

Reply:
The data consistency can happen if the developer messes with the data somehow, but that can happen with anything related to auth, then he's on his own.

We are making sure we save objects with the precise shape (or at least using Typescript telling us if we are not doing it) so there shouldn't be any surprises here.